dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts

```diff
@@ -0,0 +1,353 @@
+    this.#windowMessageTransport.connect()
+
+    const currencyIds = [
+      this.#network === BitcoinNetwork.Mainnet ? "bitcoin" : "bitcoin_testnet",
+    ]
+
+    if (!this.#hasConnectFunctionBeenCalled) {
+      const accounts: Account[] = await this.#walletApiClient.account.list({
+        currencyIds,
+      })
+
+      if (this.#options.tryConnectToAddress) {
+        for (let i = 0; i < accounts.length; i += 1) {
+          const acc = accounts[i]
+          // eslint-disable-next-line no-await-in-loop
+          const address = await this.#getAddress(acc.id)
+
+          if (address === this.#options.tryConnectToAddress) {
+            this.#account = acc
+            break
+          }
+        }
+      } else if (accounts.length === 1) {
+        ;[this.#account] = accounts
+      }
+    }
+
+    if (!this.#account || this.#hasConnectFunctionBeenCalled) {
+      this.#account = await tryRequest<Account>()(async () =>
+        this.#walletApiClient.account.request({
+          currencyIds,
+        }),
+      )
+    }
+
+    this.#hasConnectFunctionBeenCalled = true
+    return this.#account.address
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  isInstalled(): boolean {
+    return true
+  }
+
+  /**
+   * Closes the connection with Ledger Live Wallet API.
+   */
+  disconnect(): Promise<void> {
+    this.#account = undefined
+    this.#windowMessageTransport.disconnect()
+
+    return Promise.resolve()
+  }
+
+  /**
+   * In the Ledger Live Wallet API the address is "renewed" each time funds are
+   * received in order to allow some privacy. But to get the same depositor
+   * owner Ethereum address we must always get the same Bitcoin address (under
+   * the `m/purpose'/0'/accountId'/0/0` derivation path)
+   *
+   * @returns Always the same bitcoin address even the address has been
+   *          "renewed" by the Ledger Live Wallet API.
+   */
+  async getAddress(): Promise<string> {
+    if (this.#account === undefined) {
+      throw new Error("Connect first")
+    }
+
+    return this.#getAddress(this.#account.id)
+  }
+
+  #getAddress(accountId: string, derivationPath: string = "0/0") {
+    return this.#walletApiClient.bitcoin.getAddress(accountId, derivationPath)
+  }
+
+  /**
+   * Signs withdraw message.
+   * @param message Message to sign.
+   * @param data Withdrawal transaction data.
+   * @returns Hash of the signed message.
+   */
+  async signWithdrawMessage(
+    message: string,
+    data: Omit<AcreWithdrawalData, "operation" | "nonce"> & {
+      operation: number
+      nonce: number
+    },
+  ) {
+    return this.#signMessage({
+      type: AcreMessageType.Withdraw,
+      message: {
+        ...data,
+        operation: data.operation.toString(),
+        nonce: data.nonce.toString(),
+      },
+    })
+  }
+
+  /**
+   * Signs message.
+   * @param message Message to sign.
+   * @returns Hash of the signed message.
+   */
+  async signMessage(message: string): Promise<string> {
```

Comment:
We must implement the `signMessage` function from the `OrangeKitBitcoinWalletProvider`.

Reply:
Ok, so `signMessage` from `OrangeKitBitcoinWalletProvider` is a general blind signing function, the Acre module supports only clear signing.
Let's implement `signMessage` to use `walletApiClient.message.sign` to match the `OrangeKitBitcoinWalletProvider` interface.
And then we have two clear signing functions for the custom Acre module: `signSignInMessage` and `signWithdrawMessage`.

Reply:
> Let's implement `signMessage` to use `walletApiClient.message.sign` to match the `OrangeKitBitcoinWalletProvider` interface.

The `walletApiClient.message.sign` always signs the message using the fresh address so this may be confusing. Another thing is that we use the `useSignMessage` hook from `wagmi` which calls the `signMessage` function from the `bitcoinProvider` see https://github.com/thesis/orangekit/blob/main/typescript/react/src/wallet/client.ts#L32. Since the Acre dapp currently requires 2 types of messages: SIWW message (it's just string) and withdraw, and we have the custom bitcoin provider used only in Acre dapp I assumed the `signMessage` signs the message using the Acre module. Also, with the current implementation of Acre module, we can sign any message using the Acre module and `SignIn` type.

Reply:
You raised important points. We shouldn't block this PR to solve this matter since it requires work on the OrangeKit and Ledger API sides.